### PR TITLE
CLOUDP-178754: Deletion Protection for Data Federation

### PR DESCRIPTION
**Pull request by roothorp**
### All Submissions:

* [x] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [ ] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if there is one).
* [ ] Update docs/release-notes/release-notes.md if your changes should be included in the release notes for the next release.

### CODE REVIEW

**Comment by roothorp** [comment]
Based as best I could off @helderjs' https://github.com/mongodb/mongodb-atlas-kubernetes/pull/1028 - please let me know if I missed anything.

File: pkg/controller/atlasdatafederation/datafederation_controller.go [hunk 1]
```diff
@@ -110,6 +109,35 @@ func (r *AtlasDataFederationReconciler) Reconcile(contextInt context.Context, re
 	}
 	ctx.Client = atlasClient
 
+	owner, err := customresource.IsOwner(dataFederation, r.ObjectDeletionProtection, customresource.IsResourceManagedByOperator, managedByAtlas(contextInt, atlasClient))
```
Comment (josvazg) [inline pkg/controller/atlasdatafederation/datafederation_controller.go:112]: My 2 cents, we could try to increase our confidence in code with more unit tests. For this particular case this piece of code could be on a separate method that we could try to test in isolation. For an example check out[ PR #1036 protection tests](https://github.com/mongodb/mongodb-atlas-kubernetes/pull/1036/files#diff-0e68b721c4dc0a2ce82467624c0d6842d6f91467b6c6bc95e1dc414e13f6c1bf)

Not a blocker anyway, just a suggestion

**Review by josvazg — APPROVED** [review]

File: pkg/controller/atlasdatafederation/datafederation_controller.go (hunk 1 above)
Comment (roothorp) [reply pkg/controller/atlasdatafederation/datafederation_controller.go:112]: I did see the unit tests you added when I was looking at your PR last week, and I'm definitely inspired. Will look to add some today!

File: pkg/controller/atlasdatafederation/datafederation_controller.go [hunk 2]
```diff
@@ -110,6 +109,35 @@ func (r *AtlasDataFederationReconciler) Reconcile(contextInt context.Context, re
...
+	}
+
+	err = customresource.ApplyLastConfigApplied(contextInt, project, r.Client)
+	if err != nil {
+		result = workflow.Terminate(workflow.Internal, err.Error())
+		ctx.SetConditionFromResult(status.DataFederationReadyType, result)
+		log.Error(result.GetMessage())
+
+		return result.ReconcileResult(), nil
+	}
```
Comment (josvazg) [inline pkg/controller/atlasdatafederation/datafederation_controller.go:139]: Maybe it is a bit early for this piece here, I mean it should probably be set once nothing else is expected to mutate the k8s object, and I see a lot of logic below that could potentially be making changes.

Should we move this below, maybe after deletion?

File: pkg/controller/atlasdatafederation/datafederation_controller.go [hunk 3]
```diff
@@ -260,3 +263,31 @@ func (r *AtlasDataFederationReconciler) Delete(e event.DeleteEvent) error {
 
 	return nil
 }
+
+func managedByAtlas(ctx context.Context, atlasClient mongodbatlas.Client, projectID string, log *zap.SugaredLogger) customresource.AtlasChecker {
```
Comment (helderjs) [inline pkg/controller/atlasdatafederation/datafederation_controller.go:267]: The same project logic won't work here. linter is already complaining about the unused `projectID` param.

File: pkg/controller/atlasdatafederation/datafederation_controller.go [hunk 4]
```diff
@@ -260,3 +263,31 @@ func (r *AtlasDataFederationReconciler) Delete(e event.DeleteEvent) error {
...
+	return func(resource mdbv1.AtlasCustomResource) (bool, error) {
+		dataFederation, ok := resource.(*mdbv1.AtlasDataFederation)
+		if !ok {
+			return false, errors.New("failed to match resource type as AtlasDataFederation")
+		}
+
+		if dataFederation.Spec.Name == "" {
```
Comment (helderjs) [inline pkg/controller/atlasdatafederation/datafederation_controller.go:274]: This is unlikely to happen once it's a DataFederation can't exist without a name, but if you want to keep this check, better to return an error

File: pkg/controller/atlasdatafederation/datafederation_controller.go [hunk 5]
```diff
@@ -260,3 +263,31 @@ func (r *AtlasDataFederationReconciler) Delete(e event.DeleteEvent) error {
...
+			if errors.As(err, &apiError) && (apiError.ErrorCode == atlas.NotInGroup || apiError.ErrorCode == atlas.ResourceNotFound) {
+				return false, nil
+			}
+			return false, err
+		}
+
+		isSame, err := dataFederationMatchesSpec(log, atlasDataFederation, dataFederation)
+		if err != nil {
+			return true, nil
+		}
```
Comment (helderjs) [inline pkg/controller/atlasdatafederation/datafederation_controller.go:290]:
```suggestion
		}
		
```

**Review by helderjs — CHANGES_REQUESTED** [review]

File: pkg/controller/atlasdatafederation/datafederation_controller.go [hunk 6]
```diff
@@ -271,11 +271,7 @@ func managedByAtlas(ctx context.Context, atlasClient mongodbatlas.Client, projec
 			return false, errors.New("failed to match resource type as AtlasDataFederation")
 		}
 
-		if dataFederation.Spec.Name == "" {
-			return false, nil
-		}
-
-		atlasDataFederation, _, err := atlasClient.DataFederation.Get(ctx, dataFederation.Spec.Project.Name, dataFederation.Name)
+		atlasDataFederation, _, err := atlasClient.DataFederation.Get(ctx, projectID, dataFederation.Name)
```
Comment (helderjs) [inline pkg/controller/atlasdatafederation/datafederation_controller.go:274]: Just noticed, You're using k8s name. it should be `dataFederation.Spec.Name`

File: test/int/datafederation_protect_test.go [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+				df := &mongodbatlas.DataFederationInstance{
+					Name: testDataFederationName,
+				}
+
+				_, _, err := atlasClient.DataFederation.Create(context.TODO(), testProject.ID(), df)
+				Expect(err).To(BeNil())
+				Eventually(func(g Gomega) {
+					atlasDataFederation, _, err := atlasClient.DataFederation.Get(context.TODO(), testProject.ID(), testDataFederationName)
+					g.Expect(err).To(BeNil())
+					g.Expect(atlasDataFederation).ToNot(BeNil())
```
Comment (helderjs) [inline test/int/datafederation_protect_test.go:127]:
```suggestion
					g.Expect(atlasDataFederation).ToNot(BeNil())
                                        g.Expect(atlasDataFederation.State).Should(Equal("ACTIVE"))
```

**Review by helderjs — CHANGES_REQUESTED** [review]

**Review by helderjs — CHANGES_REQUESTED** [review]
All good, there's a last comment related to some refactoring done.

File: pkg/controller/atlasdatafederation/datafederation_controller.go [hunk 8]
```diff
@@ -206,22 +221,10 @@ func (r *AtlasDataFederationReconciler) readProjectResource(ctx context.Context,
...
-	err = c.Watch(&source.Kind{Type: &mdbv1.AtlasDataFederation{}}, &handler.EnqueueRequestForObject{}, r.GlobalPredicates...)
-	if err != nil {
-		return err
-	}
-
-	return nil
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("AtlasDataFederation").
+		For(&mdbv1.AtlasDataFederation{}, builder.WithPredicates(r.GlobalPredicates...)).
+		Complete(r)
```
[the remark this answers is not in the document]
Comment (roothorp) [reply pkg/controller/atlasdatafederation/datafederation_controller.go]: I think we can use `Watches()` which looks to be the same as `For()`, but you can specify the event.

File: pkg/controller/atlasdatafederation/datafederation_controller.go [hunk 9]
```diff
@@ -206,22 +222,11 @@ func (r *AtlasDataFederationReconciler) readProjectResource(ctx context.Context,
...
-	}
-	err = c.Watch(&source.Kind{Type: &mdbv1.AtlasDataFederation{}}, &handler.EnqueueRequestForObject{}, r.GlobalPredicates...)
-	if err != nil {
-		return err
-	}
-
-	return nil
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("AtlasDataFederation").
+		Watches(&source.Kind{Type: &mdbv1.AtlasDataFederation{}}, &watch.EventHandlerWithDelete{Controller: r}, builder.WithPredicates(r.GlobalPredicates...)).
```
Comment (helderjs) [inline pkg/controller/atlasdatafederation/datafederation_controller.go:227]: This looks very strange to me, but it does the same as before using the builder interface. We should visit this later.

**Review by helderjs — APPROVED** [review]

**Pull request merged**